include/seastar/core/abort_source.hh

```diff
@@ -100,15 +100,16 @@ public:
 
 private:
     using subscription_list_type = bi::list<subscription, bi::constant_time_size<false>>;
-    std::optional<subscription_list_type> _subscriptions = subscription_list_type();
+    subscription_list_type _subscriptions;
     std::exception_ptr _ex;
 
     void do_request_abort(std::optional<std::exception_ptr> ex) noexcept {
-        assert(_subscriptions);
+        if (_ex) {
+            return;
+        }
```

Comment:
let's assert that _ex evaluates to true after it's assigned below. Theoretically, either the ex param or the default ex could contain a null exception_ptr
Alternatively, the abort source could hold an optional exception and check if it's engaged here.